crates/ruff_db/src/system/memory_fs.rs

```diff
@@ -237,6 +238,24 @@ impl MemoryFileSystem {
         let normalized = SystemPath::absolute(path, &self.inner.cwd);
         normalized.into_utf8_path_buf()
     }
+
+    pub(crate) fn read_dir(&self, path: impl AsRef<SystemPath>) -> Result<DirectoryIterator> {
+        let by_path = self.inner.by_path.read().unwrap();
+        let normalized = self.normalize_path(path.as_ref());
+        let entry = by_path.get(&normalized).ok_or_else(not_found)?;
+        if entry.is_file() {
+            return Err(not_a_directory());
+        };
+        // Is this inefficient? Yes, but this implementation of `System` is just for testing,
+        // and the lifetime issues get somewhat complicated without the cloning.
+        let contents = by_path.keys().cloned().collect();
+        Ok(DirectoryIterator {
```

Comment:
Why does this copy all keys and not just collect the relevant paths?

If we already collect, then let's just return a vec iterator

Reply:
I tried for quite a while to figure out a way of doing this without collecting, but couldn't get it past the borrow checker. In the end I gave up and figured it didn't matter much, since I believed this implementation was solely for testing. But you pointed out elsewhere that that's incorrect :( so I'll revisit this.

Reply:
I don't mind collecting to a `Vec` because returning an `Iterator` that holds on to the lock can cause a dead lock if the iterating code calls another memory file system method. 

I think you can do something like this:

```rust
		let mut entries = Vec::new();

        for (entry_path, entry) in by_path.range(normalized.clone()..).skip(1) {
            if !entry_path.starts_with(&normalized) {
                break;
            }

            if entry_path.parent() == Some(&normalized) {
                let file_type = match entry {
                    Entry::File(_) => FileType::File,
                    Entry::Directory(_) => FileType::Directory,
                };

                entries.push(Ok(DirectoryEntry {
                    path: SystemPathBuf::from_utf8_path_buf(entry_path.to_path_buf()),
                    file_type,
                }));
            }
        }

        Ok(ReadDirectory {
            entries: entries.into_iter(),
        })
```

Which only collects the relevant paths instead of all of them.